staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -27,18 +27,29 @@ import (
 	resourceapi "k8s.io/api/resource/v1alpha3"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apiserver/pkg/cel/environment"
-	resourcelisters "k8s.io/client-go/listers/resource/v1alpha3"
 	"k8s.io/dynamic-resource-allocation/cel"
 	"k8s.io/klog/v2"
 )
 
-// ClaimLister returns a subset of the claims that a
+// resourceClaimLister returns a subset of the claims that a
 // resourcelisters.ResourceClaimLister would return.
-type ClaimLister interface {
+type resourceClaimLister interface {
 	// ListAllAllocated returns only claims which are allocated.
 	ListAllAllocated() ([]*resourceapi.ResourceClaim, error)
 }
 
+type resourceSliceLister interface {
+	// List returns a list of all ResourceSlices.
+	List() ([]*resourceapi.ResourceSlice, error)
+}
```

Comment:
In https://github.com/kubernetes/kubernetes/pull/127277/files#diff-8907e59dbe7c433690f8025b6cc248893a60478f6e751d4679bf0e64aa4471e7, I am completely removing this "list all slices" step. Instead, the caller needs to track allocated devices, which turned out to be more efficient because tracking can be incremental instead of always starting from scratch.

I'm on the edge whether I should ask you to hold this PR until mine is reviewed and merged, or whether I should let you go ahead and then update mine. But I would need to change the interfaces that you are implementing in CA, so perhaps it is better to get my changes in first so that you don't need to rewrite the CA code?

https://github.com/kubernetes/kubernetes/pull/127277 is blocked by https://github.com/kubernetes/kubernetes/pull/127266. Perhaps I should rebase #127277 on top of master so that we can merge it sooner? I'd prefer not to, but it would be possible.

Reply:
I forgot to mention that apart from these logistics the proposed changes look okay to me.

Reply:
I rebased https://github.com/kubernetes/kubernetes/pull/127277. Perhaps someone from SIG Scheduling can review the new commits in that PR (i.e. ignore the DRA API changes)? Then when those API changes are merged, #127277 could be merged quickly (rebase, re-add LGTM).

Reply:
@pohly I would suggest you land your PR changes first and we can adapt this one on top of it. @towca is on vacation for another week or so, so you can take the lead position.

If https://github.com/kubernetes/kubernetes/pull/127266 and https://github.com/kubernetes/kubernetes/pull/127277 land early next week, then SIG Autoscaling could spawn a new PR for this CA integration from this one, incorporating your feedback and the changes necessitated by your new DRA foo in master.

cc @MaciekPytel @nojnhuh @ritazh

Reply:
#127266 might land early next week, but for #127277 I still need a reviewer from SIG scheduling for the final LGTM and approval. It was reviewed already by a member of WG Device management. Any volunteers?

Reply:
cc @alculquicondor @Huang-Wei @ahg-g @kerthcet @sanposhiho

Reply:
Hmm, did you mean removing the "list all allocated claims" step? From a quick read of #127277, the allocator still needs all ResourceSlices + all allocated devices from allocated ResourceClaims, right?

So `framework.ResourceClaimLister` can stay as proposed, but in `framework.ResourceClaimTracker` we'd need to replace `ListAllAllocated()` with something like `ListAllAllocatedDevices()`? If so, that's a trivial change in CA code - the first implementation can just create the list from scratch every time. So I'm okay with merging in either order, but we don't have much time.

Reply:
Yes, list all allocated claims `ListAllAllocated()` need to be replaced by `ListAllAllocatedDevices()`, but it will be still backed by assume cache, so I believe the interface can remain the same.

The change that #127277 brings is that the device list (previously claim list) is snapshotted in PreFilter and used in Filter as static data, so the device lister (previously claim lister) does not need to be passed here. The same applies to resource slice lister. The class lister will be the only remaining one that needs to be passed.

Reply:
Done